devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
+
+Consumers of `job_queue` must create a "`session`" before they can queue requests. A session can make two kinds of requests:
```

Comment:
>Consumers [](http://example.com/codeflow?start=0&length=9)

Is this essentially a "multi-producer multi-consumer queue" (since we currently have a multi-producer single-consumer lockless queue)?

Reply:
I'm not very familiar with this terminology, but I would think that "single-producer multi-consumer queue" would be a better description. 

I intend to be pushing jobs (calling `job_queue_push`) from the same thread.

In GeoReplication terms, each job will be a `BeginReplicate` call, so these must be in the order of LSN. Each state provider shall use a separate `job_queue`.